File: src/main/java/org/cerion/stocklist/indicators/AverageTrueRange.java
package org.cerion.stocklist.indicators;

import org.cerion.stocklist.PriceList;
import org.cerion.stocklist.arrays.FloatArray;
import org.cerion.stocklist.functions.types.Indicator;

public class AverageTrueRange extends IndicatorBase {

    public AverageTrueRange() {
        super(Indicator.ATR, 14);
    }

    public AverageTrueRange(Number ...params) {
        this();
        setParams(params);
    }

    @Override
    public String getName() {
        return "Average True Range";
    }

    @Override
    public FloatArray eval(PriceList list) {
        return averageTrueRange(list, getInt(0) );
    }

    private static FloatArray averageTrueRange(PriceList list, int period)
    {
        FloatArray result = new FloatArray(list.size());

        //Current ATR = [(Prior ATR x 13) + Current TR] / 14
        result.mVal[0] = list.tr(0);
        for(int i = 1; i < list.size(); i++)
            result.mVal[i] = ((result.get(i-1) * (period-1)) + list.tr(i)) / period;

        return result;
    }
}

File: src/main/java/org/cerion/stocklist/indicators/AverageTrueRange.kt
package org.cerion.stocklist.indicators

import org.cerion.stocklist.PriceList
import org.cerion.stocklist.arrays.FloatArray
import org.cerion.stocklist.functions.types.Indicator

class AverageTrueRange(period: Int = 14) : IndicatorBase(Indicator.ATR, period) {

    override fun getName(): String {
        return "Average True Range"
    }

    override fun eval(list: PriceList): FloatArray {
        return averageTrueRange(list, getInt(0))
    }

    private fun averageTrueRange(list: PriceList, period: Int): FloatArray {
        val result = FloatArray(list.size)

        //Current ATR = [(Prior ATR x 13) + Current TR] / 14
        result.mVal[0] = list.tr(0)
        for (i in 1 until list.size)
            result.mVal[i] = (result.get(i - 1) * (period - 1) + list.tr(i)) / period

        return result
    }
}
